### ISSUE TITLE
refactor(facade): change EventEmitter to match RxJS-Next Subject (and rename plz)

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#4149]
Since EventEmitter is essentially a subject (implements `Observable` and `Observer`), it should just alias to `Subject` for JS, and match the structure and semantics for Dart (note: it should be validated that `Subject` semantics can be mapped to Dart).

Also, we should rename `EventEmitter`, since it's often confused with node's `EventEmitter`. What about just calling it Subject?

CC: @robwormald @vsavkin @blesh

**Comment by jeffbcross** [comment]
Chatting with @blesh some, he brings up that we should really evaluate if/why we need `EventEmitter` instead of just `Observable`, and if `Subject` is really the best abstraction for it. For one, a `Subject` cannot be re-subscribed to once disposed.

**Comment by robwormald** [comment]
Agreed on the above. I reckon we should strive where possible to use Observables over Subjects - this will keep things naturally cold. This also means educating users on the appropriate way to wrap things with Observables. 

The one place the name EventEmitter currently makes sense is well, emitting events from components for external consumption, but not so much elsewhere (FormBuilder, Http). Are there any other (common) cases where a user will actually want to manually "emit"? 

What about defining a specific Type for the actual EventEmitter use case (something like `NgEventBroadcaster` - don't love that name), or `EventDispatcher` and using it only for component events? Alternately, we could just use the actual EventEmitter semantics ( .on / .emit ) and leverage Rx's Observable.fromEvent? This would be closer to vanilla javascript and make interop with other things nicer.

**Comment by blesh** [comment]
Subjects are also heavier to use than Observables. They've got additional function references, an underlying array of observers, etc.

**Comment by jeffbcross** [comment]
IRL discussion recap:

 * Every place inside core that's currently using `EventEmitter` should instead use `Observable`
 * We should rename `EventEmitter`, but not sure to what. Something that doesn't confuse with Node's `EventEmitter`, and something that doesn't require understanding Rx `Subject`.
 * `EventEmitter` should just be an `Observable`, but with an extra method called `emit`, since it is only used for emitting events. `throw` and `complete` methods not necessary. This is to provide a clean API, and more memory-efficient means of for the specific use case of emitting events in custom components.

Should emitting values be Sync or Async?
 * Default sync. Because encourages good change detection behavior, since will most likely throw an error if emitting during change detection.
 * One problem with sync: Dart. Should look how Dart handles DOM events with streams, and see if there's a way to synchronously emit.

**Comment by kasperpeulen** [comment]
Coming from Dart, I would probably call `EventEmitter` just `EventStream`. Because I believe that is what it is in Dart. Not sure though how the typescript version is implemented.

**Comment by kasperpeulen** [comment]
@jeffbcross 

I'm not an expert, but I think if you want to make EventEmitter sync by default in dart, you should make it like this: 

```dart
class EventEmitter extends Stream {
  StreamController<dynamic> _controller;

  EventEmitter({sync: true}) {
    _controller = new StreamController.broadcast(sync: true);
  }

  StreamSubscription listen(void onData(dynamic line),
      {void onError(Error error), void onDone(), bool cancelOnError}) {
    return _controller.stream.listen(onData,
        onError: onError, onDone: onDone, cancelOnError: cancelOnError);
  }

  void add(value) {
    _controller.add(value);
  }

  void addError(error) {
    _controller.addError(error);
  }

  void close() {
    _controller.close();
  }
}
```

**Comment by blesh** [comment]
You need a Dart Observable, IMO. It would be lighter weight than EventStream, but compatible with it. In reality, EventStream looks like it could be subclassed from an Observable.

**Comment by IgorMinar** [comment]
bumping to a40 because no breaking changes in a39, but a40 is the last chance to make this before AC.

**Comment by naomiblack** [comment]
this should be resolved by #4604

**Comment by jeffbcross** [comment]
Moving this into 41, working on a simple solution to manually add the combinators we need to EventEmitter without changing anything about its inheritance.
